<?xml version="1.0"?>

<!-- A QUADRUPED NAMED TACHYON. I just like that name :) -->

<robot name="tachyon" xmlns:xacro="http://www.ros.org/wiki/xacro">

<!-- Lower and upper limits for the shoulder joint (base_link_x_y_leg) -->
<xacro:property name="lower_shoulder" value="-0.90"/>
<xacro:property name="upper_shoulder" value="0.90"/>

<xacro:property name="damp" value="0.3"/>
<xacro:property name="friction" value="0.1"/>

<xacro:include filename="$(find tachyon)/urdf/leg.xacro"/>
<xacro:include filename="$(find tachyon)/urdf/tachyon.gazebo"/>

<!-- Front and back legs -->
<xacro:leg limb_name="front_left" lower_upper_limb="-2.06" upper_upper_limb="2.06" lower_lower_limb="-2.20" upper_lower_limb="2.20"/>
<xacro:leg limb_name="front_right" lower_upper_limb="-2.06" upper_upper_limb="2.06" lower_lower_limb="-2.20" upper_lower_limb="2.20"/>
<xacro:leg limb_name="back_left" lower_upper_limb="-2.06" upper_upper_limb="2.06" lower_lower_limb="-2.20" upper_lower_limb="2.20"/>
<xacro:leg limb_name="back_right" lower_upper_limb="-2.06" upper_upper_limb="2.06" lower_lower_limb="-2.20" upper_lower_limb="2.20"/>

<xacro:macro name="default_inertial" params="z_value i_value mass">
    <inertial>
        <origin xyz="0 0 ${z_value}" rpy="0 0 0"/>
        <mass value="${mass}"/>
        <inertia ixx="${i_value}" ixy="0" ixz="0" iyy="${i_value}" iyz="0" izz="${i_value}"/>
    </inertial>
</xacro:macro>

<!-- Considering that model files are .STL, material is rather redundant *shrugs* -->
<xacro:macro name="sw_default_material">
    <material name="">
        <color rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
    </material>
</xacro:macro>    

<!-- Base Link -->
<link name="base_link">
    <visual>
        <origin xyz="0 0 1" rpy="0 0 0"/>
        <geometry>
            <mesh filename="package://tachyon/meshes/base_link.STL" />
        </geometry>   
        <material name="">
        <color rgba="0 0 0 1" />
      </material>
    </visual>

    <collision>
        <origin xyz="0 0 1" rpy="0 0 0"/>
        <geometry>
            <box size="1 0.4 0.1"/>
        </geometry>        
    </collision>

    <inertial>
        <origin
          xyz="-3.1155E-17 5.5984E-17 1"
          rpy="0 0 0" />
        <mass
          value="10" />
        <inertia
          ixx="0.14"
          ixy="8.3579E-18"
          ixz="1.1515E-16"
          izz="0.95"
          iyz="2.2791E-21"
          iyy="0.83" />
      </inertial>

</link>


<!-- Front Marker: to indicate the front of tachyon -->
<link name="front_marker">
    <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <box size="0.03 0.2 0.03"/>
        </geometry>   
        <material name="">
        <color rgba="1 1 1 1" />
      </material>
    </visual>

    <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <box size="0.03 0.2 0.03"/>
        </geometry>        
    </collision>
</link>

<joint name="base_link_to_front_marker" type="fixed">
    <parent link="base_link"/>
    <child link="front_marker"/>
    <origin xyz="0.51 0 1.03" rpy="0 0 0"/>  
</joint>

<joint name="base_link_to_front_left_leg" type="revolute">
    <parent link="base_link"/>
    <child link="swing_link_front_left"/>
    <origin xyz="0.47 0.25825 1" rpy="0 0 3.14"/>
    <limit effort="1000" lower="${lower_shoulder}" upper="${upper_shoulder}" velocity="0.5"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="${damp}" friction="${friction}" />    
</joint>

<joint name="base_link_to_front_right_leg" type="revolute">
    <parent link="base_link"/>
    <child link="swing_link_front_right"/>
    <origin xyz="0.47 -0.25825 1" rpy="0 0 0"/>
    <limit effort="1000" lower="${lower_shoulder}" upper="${upper_shoulder}" velocity="0.5"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="${damp}" friction="${friction}" /> 
</joint>

<joint name="base_link_to_back_left_leg" type="revolute">
    <parent link="base_link"/>
    <child link="swing_link_back_left"/>
    <origin xyz="-0.47 0.25825 1" rpy="0 0 3.14"/>
    <limit effort="1000" lower="${lower_shoulder}" upper="${upper_shoulder}" velocity="0.5"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="${damp}" friction="${friction}" /> 
</joint>

<joint name="base_link_to_back_right_leg" type="revolute">
    <parent link="base_link"/>
    <child link="swing_link_back_right"/>
    <origin xyz="-0.47 -0.25825 1" rpy="0 0 0"/>
    <limit effort="1000" lower="${lower_shoulder}" upper="${upper_shoulder}" velocity="0.5"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="${damp}" friction="${friction}" /> 
</joint>

<xacro:default_transmission name="base_to_front_left" joint_name="base_link_to_front_left_leg"/>

<xacro:default_transmission name="base_to_front_right" joint_name="base_link_to_front_right_leg"/>

<xacro:default_transmission name="base_to_back_left" joint_name="base_link_to_back_left_leg"/>

<xacro:default_transmission name="base_to_back_right" joint_name="base_link_to_back_right_leg"/>

</robot>